cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md

```diff
@@ -346,6 +346,44 @@ the device view can be obtained via function `column_device_view::create(column_
 data, a specialized device view for list columns can be constructed via
 `lists_column_device_view(column_device_view)`.
 
+# libcudf policies and design principles
+
+Some aspects of libcudf usage may be surprising to new users.
+As a performance-oriented library, many design decisions prioritize flexibility and performance in ways that require some additional work or input from client code.
+We document these policies and the reasons behind them here:
+
+## libcudf does not introspect data
+
+libcudf APIs generally do not perform deep introspection and validation of input data.
+There are numerous reasons for this:
+1. It violates the single responsibility principle: validation is separate from execution.
+2. Since libcudf data structures store data on the GPU, any validation incurs the minimal overhead of a kernel launch.
+3. API promises around data introspection often significantly complicate implementation.
+
+Users are therefore responsible for passing valid data into such APIs.
+_Note that this policy does not mean that libcudf performs no validation whatsoever_.
+libcudf APIs should still perform any validation that does not require introspection, such as checking whether input column/table sizes or dtypes are appropriate.
```

Comment:
This extends to output sizes and results as well.

In general there is no validation that an operation will produce a column that fits in the 2GB limit. There are a few minor exceptions where the validation does not require an extra kernel launch.

Also, operations that may produce overflow (integer or floating-point arithmetic) are also not checked. Overflow produces undefined results. For example, integer overflow to negative values is not guaranteed.

Reply:
My intent here was not to provide exhaustive lists of things we will and will not check, but rather to provide some examples of cases to illustrate the difference between introspective and non-introspective operations. I am wary of trying to maintain a complete list that will inevitably go out of date the day after it is published. Do you think it's worth trying to keep a larger list?

Reply:
I suppose I would like to keep a list of things that keep being brought up (like these two here) so that I can point to this document to say this is by design.

Reply:
I agree this does not need to be _exhaustive_, but it would be nice if this policy guide can serve as a mini-FAQ of "why is libcudf designed this way?" I think that the concrete examples @davidwendt mentioned help illustrate the philosophy of the library.

Reply:
I've added some, and I've reworded this section to reflect that we want to keep a list but make no promises that it's exhaustive.